### PR TITLE
block log pruning: optionally reduce disk usage of block log file during operation by periodically trimming blocks

### PR DESCRIPTION
**Pull request by spoonincode**
Add a new option, `block-log-prune-blocks`, which when set will periodically reduce the disk usage of the block log to only that of the given number of most recent blocks. This allows operators to limit the disk usage of the block log without stopping and restarting nodeos.

Reducing disk usage is performed by "poking holes" in the log file; in other words making it a sparse file. This may have implications if the pruned log is tarballed/copied/etc with an application that does not recognize sparse files: the log may take up more disk space than expected on its destination if an application is not spare file aware.

A new command line option `--vacuum` is added to `eosio-blocklog` to squash the log back to a simple layout thus removing the spare file nature of the file. This vacuum operation is also automatically performed at nodeos exit if the disk usage of the block log is less than 1GB. (feedback welcome on if this size should be configurable etc)

Pruning is only periodically performed so in practice a small number of additional blocks (no more than a few megabytes worth of data) will be available beyond the configured number.

### Code review notes

After trying a couple other approaches I settled on this one to reduce the amount of invasive and/or complex changes to the current log implementation. That said, there still ended up being more changes than I was expecting.

Besides the previously mentioned concern that a sparse file could have its holes "lost" during a transfer, it is notable that there is no way to perform a log recovery because if the end of the log is damaged (which is what would cause need to  perform a recovery) the trailer on the log storing how many blocks are present must also be assumed damaged. This potentially could be resolved by always placing and writing to a 'tail' indicator at the front of the log every time its appended to. However the block log doesn't have a lot of space in its header to commandeer for this purpose. Feedback welcome on how problematic this is.

There are minor improvements that could be made that didn't seem worth it. For example, the index doesn't get holes poked in it but it's just such a small file anyways.

### CODE REVIEW

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -42,11 +42,14 @@ namespace eosio { namespace chain {
             signed_block_ptr         head;
             block_id_type            head_id;
             fc::cfile                block_file;
+            size_t                   log_blk_size;
```
Comment (spoonincode) [inline libraries/chain/block_log.cpp:45]: unused code that needs to be removed

**Comment by spoonincode** [comment]
Some tasks to carry over from ship log review:
* double check if any preconditions on functions on added prune/vacuum functions should be more explicit
* bundle all the various pruning options in a struct instead of making some assumptions various places
* can probably optimize the hole punching in vacuum better

File: libraries/chain/block_log.cpp [hunk 2]
```diff
@@ -453,13 +675,20 @@ namespace eosio { namespace chain {
       if (my->block_file.tellp() <= sizeof(pos))
          return {};
 
+      //figure out if this is a pruned log or not. we can't just look at the configuration since
+      // read_head() is called early on, and this isn't hot enough to warrant a member bool to track it
+      my->block_file.seek(0);
+      uint32_t current_version;
+      fc::raw::unpack(my->block_file, current_version);
+      bool is_currently_pruned = detail::is_pruned_log_and_mask_version(current_version);
```
Comment (brianjohnson5972) [inline libraries/chain/block_log.cpp:683]: make const

File: libraries/chain/block_log.cpp [hunk 3]
```diff
@@ -775,6 +1019,7 @@ namespace eosio { namespace chain {
       _version = 0;
       auto size = fread((char*)&_version, sizeof(_version), 1, _file.get());
       EOS_ASSERT( size == 1, block_log_exception, "Block log file at '${blocks_log}' could not be read.", ("file", _block_file_name) );
+      bool is_prune_log = is_pruned_log_and_mask_version(_version);
```
Comment (brianjohnson5972) [inline libraries/chain/block_log.cpp:1022]: make const

File: libraries/chain/block_log.cpp [hunk 4]
```diff
@@ -788,8 +1033,18 @@ namespace eosio { namespace chain {
       auto status = fseek(_file.get(), 0, SEEK_END);
       EOS_ASSERT( status == 0, block_log_exception, "Could not open Block log file at '${blocks_log}'. Returned status: ${status}", ("blocks_log", _block_file_name)("status", status) );
 
-      _eof_position_in_file = ftell(_file.get());
+      auto _eof_position_in_file = ftell(_file.get());
```
Comment (brianjohnson5972) [inline libraries/chain/block_log.cpp:1036]: rename eof_position_in_file

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]
Some of these have to be handled for this PR to merge, some can be put in a github issue to be handled after a release and some can be disputed.

File: libraries/chain/block_log.cpp [hunk 5]
```diff
@@ -56,6 +58,18 @@ namespace eosio { namespace chain {
             void reopen();
 
             void close() {
+               //for a pruned log that has at least one block, see if we should vacuum it
+               //pruned_blocks == UINT32_MAX is kind of a hint to "please don't prune, but keep it a pruned log"
+               if( block_file.is_open() && index_file.is_open() && head && prune_blocks && *prune_blocks != UINT32_MAX) {
+                  const size_t first_data_pos = get_block_pos(first_block_num);
+                  block_file.seek_end(-sizeof(uint32_t));
+                  const size_t last_data_pos = block_file.tellp();
+                  if(last_data_pos - first_data_pos < 1024*1024*1024) {
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: It was a nice round number to start discussion at. It’s not clear whether there is a preference when/if vacuuming should be performed automatically. I _thought_ I had a strong opinion on this initially but all the ongoing discussions make me uncertain now.

Clearly vacuuming is performed with the manual `eosio-blocklog --vacuum` command (block log is fortunate that it has this separate tool: ship has no such operation making this decision matrix a little tougher to decide on)

For automatic vacuuming there are then two opportunities:
* On exit, if the size to vacuum is small enough (what’s small enough? 1GB at the moment).
* On start, if the user no longer specifies a trim value. (the alternative here is that the log stays an “unvacuumed pruned log” but stops performing any more pruning until the user specifies otherwise in the future).

fwiw I've removed automatic vacuuming at the moment until we can come to agreement on the criteria for when it should happen.

File: libraries/chain/block_log.cpp [hunk 6]
```diff
@@ -72,12 +86,28 @@ namespace eosio { namespace chain {
...
+
+            size_t convert_existing_header_to_vacuumed();
+
+            uint64_t get_block_pos(uint32_t block_num);
 
             template <typename ChainContext, typename Lambda>
             static std::optional<ChainContext> extract_chain_context( const fc::path& data_dir, Lambda&& lambda );
       };
 
+      constexpr uint32_t pruned_version_flag = 1<<16;
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: This was to mirror that of ship, where I more clearly split the uint32_t version in to two uint16_t. No problem moving it to be MSb though.

File: libraries/chain/block_log.cpp [hunk 7]
```diff
@@ -327,6 +425,122 @@ namespace eosio { namespace chain {
...
+      //go ahead and write a new valid header now. if the vacuum fails midway, at least this means maybe the
+      // block recovery can get through some blocks.
+      size_t copy_to_pos = convert_existing_header_to_vacuumed();
+
+      version = block_log::max_supported_version;
+      prune_blocks.reset();
+
+      //if there is no head block though, bail now, otherwise first_block_num won't actually be available
+      // and it'll mess this all up
+      if(!head)
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: Not following this comment, help me out. No head block means no entries in the log which should mean index is completely empty by my understanding. But digging on this some did show another issue and I’ve added a unit test to address prune transitions on an empty blog.

File: libraries/chain/block_log.cpp [hunk 8]
```diff
@@ -327,6 +425,122 @@ namespace eosio { namespace chain {
...
+
+      std::vector<char> buff;
+      buff.resize(4*1024*1024);
+
+      auto tick = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now());
+      while(copy_sz) {
+         const size_t copy_this_round = std::min(buff.size(), copy_sz);
+         block_file.seek(copy_from_pos);
+         block_file.read(buff.data(), copy_this_round);
+         block_file.punch_hole(copy_from_pos, copy_from_pos+copy_this_round);
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: Ensures that throughout the vacuuming procedure the disk usage of the file remains the same (plus maybe a block or two).

File: libraries/chain/block_log.cpp [hunk 9]
```diff
@@ -286,38 +344,78 @@ namespace eosio { namespace chain {
...
+         const uint64_t end = block_file.tellp();
          index_file.write((char*)&pos, sizeof(pos));
          head = b;
          head_id = b->calculate_id();
 
-         flush();
+         if(prune_blocks) {
+            //only bother to try pruning every 4MB written. except for when the prune is set very low (mainly for unit test purposes)
+            uint64_t bother_every = 4*1024*1024;
+            if(*prune_blocks < 16)
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: This has been refactored out with a proper config struct so that the behavior only exists when unit test sets it up that way.

File: libraries/chain/block_log.cpp [hunk 10]
```diff
@@ -286,38 +344,78 @@ namespace eosio { namespace chain {
...
+               bother_every = 4096;
+
+            const uint64_t mask = ~(bother_every-1);
+            if((pos&mask) != (end&mask))
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: Yeah this was confusion on the ship review too. Every time we add a block we check the start (`pos`) and `end` and see if it straddled `bother_every`. I did this so I didn’t have to have yet another member variable tracking how much written but admittedly it’s a bit unusual. fwiw `bother_every` is gone, in favor of using the proper config struct

File: libraries/chain/block_log.cpp [hunk 11]
```diff
@@ -453,13 +675,20 @@ namespace eosio { namespace chain {
...
+      // read_head() is called early on, and this isn't hot enough to warrant a member bool to track it
+      my->block_file.seek(0);
+      uint32_t current_version;
+      fc::raw::unpack(my->block_file, current_version);
+      bool is_currently_pruned = detail::is_pruned_log_and_mask_version(current_version);
+
       my->block_file.seek_end(-sizeof(pos));
-      my->block_file.read((char*)&pos, sizeof(pos));
-      if (pos != npos) {
+      if(is_currently_pruned)
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp:687]: I reordered the seeks & skips to be more in line with what is actually be skipped over in the file. I don't think the mouthful with the ternary operator improves readability and the extra seek isn't much of a big deal for this function since it isn't hot.

**Comment by brianjohnson5972** [comment]
I think <block-logs-prune-blocks> is a confusing config parameter name, since it sounds like the number of blocks that get pruned. Would like a name like <prune-block-log-retain>.

**Comment by spoonincode** [comment]
Maybe `block-log-retain-blocks`? I'd like to keep the option starting with "block-log"

File: libraries/chain/block_log.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: Once someone enables pruning the header will have the flag set that its a pruned log, so when vacuuming (i.e. converting it back to a non-pruned log) it firsts clears the flag by way of rewriting the header via `convert_existing_header_to_vacuumed()`. Then if `!head` then there are no blocks & the index is empty so we're done. Maybe the comment saying "bail" has the wrong connotation -- it's really that we're done here. Otherwise if there are blocks (head is nonnull) the magic has to happen.

[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: I agree now with your original code, but I think your new code is wrong. I don't see the convert_existing_header_to_vacuumed() adding the extra "valid non-pruned blocks count" (or whatever that name is) 32 bit number at the end, so why are you removing that 32 bits from the end?

**Comment by brianjohnson5972** [comment]
yeah, go with `block-log-retain-blocks`

File: libraries/chain/block_log.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: (I'm going to refer to the 4 byte block count on the end of the file as the _trailer_)

Creation of a new log is done in `reset()`. This can create a new log initially pruned or nonpruned. It'll write out the header with the correct flag (at the end of the function), as well as write out the trailer either by way of `append()` if there is a first block, or by writing out 0 manually if not (`head==nullptr`).

Conversion from non-pruned -> pruned occurs in `open()`. There, after calling `prune()` it _always_ writes a new header with the flag enabled and the 4 byte trailer to the file. If `head==nullptr` it'll write out a 0.

`vacuum()` is called to transition pruned -> non-pruned. The first thing it does is call `convert_existing_header_to_vacuumed()` which EOS_ASSERTs that we are indeed currently a pruned log file on disk. Then it writes a new header which will have the flag cleared. (it also may need to transition from genesis_state to chain_id). Once the header is written it'll either a) if `head==nullptr` simply chop off the trailer that was added by `reset()`/`open()`, or then if there is a `head` ergo `first_block_num` really exists then it goes on to perform the vacuum procedure which effectively eliminates the trailer.

[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/chain/block_log.cpp]: the original code had a defect where if a log was in pruned mode with no blocks and you vacummed it (i.e. converted to non pruned), that `if(!head) return` would leave the trailer on the file after clearing the flag. The log then said it was a nonpruned but the trailer was still on it which of course caused it to be a broken log. Hence adding the code to trim off the trailer there

**Review by brianjohnson5972 — APPROVED** [review]
